### ISSUE TITLE
pandas : MemoryError

### ISSUE THREAD

**Issue by PN12** [pandas-dev/pandas#24836]
#### Code Sample, a copy-pastable example if possible

```python
# Your code here

```
#### Problem description

[this should explain **why** the current behaviour is a problem and why the expected output is a better solution.]

**Note**: We receive a lot of issues on our GitHub tracker, so it is very possible that your issue has been posted before. Please check first before submitting so that we do not have to handle and close duplicates!

**Note**: Many problems can be resolved by simply upgrading `pandas` to the latest version. Before submitting, please check if that solution works for you. If possible, you may want to check if `master` addresses this issue, but that is not necessary.

For documentation-related issues, you can check the latest versions of the docs on `master` here:

https://pandas-docs.github.io/pandas-docs-travis/

If the issue has not been resolved there, go ahead and file it in the issue tracker.

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]

</details>

**Comment by PN12** [comment]
Hi,
I am getting an intermittent MemoryError while running code. The same code runs fine on re-installing Anaconda or while re-starting Jupyter Notebook / Sypder Kernel.
The system is 8GB RAM , 64bit OS.
As per the inputs received from Spyder community , this error relates to pandas. Please suggest the resolution steps.
Please note the same error occurs while running pd.read_csv command as well. 

Below is the snapshot of the same.

***********************************************************************************************
File "", line 23, in 
df_hist_trans_group = df_hist_trans.groupby ('card_id').agg(aggs)

File "C:\Users\Prachi\Anaconda3\lib\site-packages\pandas\core\groupby\groupby.py", line 4656, in aggregate
return super(DataFrameGroupBy, self).aggregate(arg, *args, **kwargs)

File "C:\Users\Prachi\Anaconda3\lib\site-packages\pandas\core\groupby\groupby.py", line 4087, in aggregate
result, how = self._aggregate(arg, _level=_level, *args, **kwargs)

File "C:\Users\Prachi\Anaconda3\lib\site-packages\pandas\core\base.py", line 490, in _aggregate
result = _agg(arg, _agg_1dim)

File "C:\Users\Prachi\Anaconda3\lib\site-packages\pandas\core\base.py", line 441, in _agg
result[fname] = func(fname, agg_how)

File "C:\Users\Prachi\Anaconda3\lib\site-packages\pandas\core\base.py", line 424, in _agg_1dim
return colg.aggregate(how, _level=(_level or 0) + 1)

File "C:\Users\Prachi\Anaconda3\lib\site-packages\pandas\core\groupby\groupby.py", line 3485, in aggregate
(_level or 0) + 1)

File "C:\Users\Prachi\Anaconda3\lib\site-packages\pandas\core\groupby\groupby.py", line 3558, in _aggregate_multiple_funcs
results[name] = obj.aggregate(func)

File "C:\Users\Prachi\Anaconda3\lib\site-packages\pandas\core\groupby\groupby.py", line 3479, in aggregate
return getattr(self, func_or_funcs)(*args, **kwargs)

File "C:\Users\Prachi\Anaconda3\lib\site-packages\pandas\core\groupby\groupby.py", line 3744, in nunique
ids, _, _ = self.grouper.group_info

File "pandas_libs\properties.pyx", line 36, in pandas._libs.properties.CachedProperty.get

File "C:\Users\Prachi\Anaconda3\lib\site-packages\pandas\core\groupby\groupby.py", line 2335, in group_info
comp_ids, obs_group_ids = self._get_compressed_labels()

File "C:\Users\Prachi\Anaconda3\lib\site-packages\pandas\core\groupby\groupby.py", line 2351, in _get_compressed_labels
all_labels = [ping.labels for ping in self.groupings]

File "C:\Users\Prachi\Anaconda3\lib\site-packages\pandas\core\groupby\groupby.py", line 2351, in 
all_labels = [ping.labels for ping in self.groupings]

File "C:\Users\Prachi\Anaconda3\lib\site-packages\pandas\core\groupby\groupby.py", line 3070, in labels
self._make_labels()

File "C:\Users\Prachi\Anaconda3\lib\site-packages\pandas\core\groupby\groupby.py", line 3103, in _make_labels
self.grouper, sort=self.sort)

File "C:\Users\Prachi\Anaconda3\lib\site-packages\pandas\util_decorators.py", line 178, in wrapper
return func(*args, **kwargs)

File "C:\Users\Prachi\Anaconda3\lib\site-packages\pandas\core\algorithms.py", line 630, in factorize
na_value=na_value)

File "C:\Users\Prachi\Anaconda3\lib\site-packages\pandas\core\algorithms.py", line 476, in _factorize_array
na_value=na_value)

File "pandas_libs\hashtable_class_helper.pxi", line 1420, in pandas._libs.hashtable.StringHashTable.get_labels

MemoryError

**Comment by gfyoung** [comment]
@PN12 : Thanks for reporting this!  In order for us to help you, could you do two things in your original issue that you posted:

* Provide output of `pandas.show_versions` ?
* Provide the code sample that causes this massive traceback?

**Comment by PN12** [comment]
@gfyoung - Please find below the details - 

**- Provide output of pandas.show_versions ?**

****Code** - pd.show_versions (as_json = False)**

**Output** - 
*******************************
pd.show_versions (as_json = False)

INSTALLED VERSIONS
------------------
commit: None
python: 3.7.1.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
byteorder: little
LC_ALL: None
LANG: en
LOCALE: None.None

pandas: 0.23.4
pytest: 4.0.2
pip: 18.1
setuptools: 40.6.3
Cython: 0.29.2
numpy: 1.15.4
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 7.2.0
sphinx: 1.8.2
patsy: 0.5.1
dateutil: 2.7.5
pytz: 2018.7
blosc: None
bottleneck: 1.2.1
tables: 3.4.4
numexpr: 2.6.8
feather: None
matplotlib: 3.0.2
openpyxl: 2.5.12
xlrd: 1.2.0
xlwt: 1.3.0
xlsxwriter: 1.1.2
lxml: 4.2.5
bs4: 4.6.3
html5lib: 1.0.1
sqlalchemy: 1.2.15
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
********************************************************
**- Provide the code sample that causes this massive traceback?**

**- The code sample is as below -**
 
Memory Error is encountered through the below commands -

1. df_hist_trans_group = df_hist_trans.groupby ('card_id').agg(aggs)

2. df_hist_trans = pd.read_csv(r'C:\Users\Prachi\Downloads\all\historical_transactions.csv')_

*************************
`# -*- coding: utf-8 -*-
"""
Created on Sat Jan 19 19:53:23 2019

@author: Prachi
"""
import numpy as np
import pandas as pd
import datetime
import gc
import matplotlib.pyplot as plt
import seaborn as sns
import lightgbm as lgb
from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import mean_squared_error
import warnings
warnings.filterwarnings('ignore')
np.random.seed(4590)

df_train = pd.read_csv(r'C:\Users\Prachi\Downloads\all\train.csv')
df_test = pd.read_csv(r'C:\Users\Prachi\Downloads\all\test.csv')

df_hist_trans = pd.read_csv(r'C:\Users\Prachi\Downloads\all\historical_transactions.csv')

df_new_merchant_trans = pd.read_csv(r'C:\Users\Prachi\Downloads\all\new_merchant_transactions.csv')

for df in [df_hist_trans,df_new_merchant_trans]:
    df['category_2'].fillna(1.0,inplace=True)
    df['category_3'].fillna('A',inplace=True)
    df['merchant_id'].fillna('M_ID_00a6ca8a8a',inplace=True)
    
def get_new_columns (name , aggs):
    return (name + '_' + k + '_' + agg 
                for k in aggs.keys() 
                    for agg in aggs[k])

for df in [df_hist_trans,df_new_merchant_trans]:
    df['purchase_date'] = pd.to_datetime(df['purchase_date'])
    df['year'] = df['purchase_date'].dt.year
    df['weekofyear'] = df['purchase_date'].dt.weekofyear
    df['month'] = df['purchase_date'].dt.month
    df['dayofweek'] = df['purchase_date'].dt.dayofweek
    df['weekend'] = (df.purchase_date.dt.weekday >=5).astype(int)
    df['hour'] = df['purchase_date'].dt.hour
    df['authorized_flag'] = df['authorized_flag'].map({'Y':1, 'N':0})
    df['category_1_n'] = df['category_1'].map({'Y': 1, 'N': 0}) 
    df['month_diff'] = ((datetime.datetime.today() - df['purchase_date']).dt.days)//30
    df['month_diff'] += df['month_lag']
    df['purchase_date_n'] = df['purchase_date'].dt.date
    df['purchase_date_n_sq'] = pd.to_datetime (df['purchase_date_n'])

#df_hist_trans['purchase_date_n']

aggs = {}

for col in ['month' , 'hour','weekofyear','dayofweek', 'year','subsector_id','merchant_id','merchant_category_id']:
    aggs[col] = ['nunique']

aggs['purchase_amount'] = ['sum','max','min','mean','var']
aggs['installments'] = ['sum','max','min','mean','var']
aggs['purchase_date_n_sq'] = ['max','min']
aggs['month_lag'] = ['max','min','mean','var']
aggs['month_diff'] = ['mean']
aggs['authorized_flag'] = ['sum', 'mean']
aggs['weekend'] = ['sum', 'mean']
aggs['category_1_n'] = ['sum', 'mean']
aggs['card_id'] = ['size']

for col in ['category_2' , 'category_3']:
    df_hist_trans[col+'_mean'] = df_hist_trans.groupby([col])['purchase_amount'].transform('mean')
    aggs[col+'_mean'] = ['mean']

new_columns = get_new_columns ('hist', aggs) 

#type(new_columns)

#list(df_hist_trans.columns)

#pd.set_option('display.max_columns', 30)
#df_hist_trans.head()

df_hist_trans['today'] = str(datetime.datetime.today()).split()[0]

df_hist_trans['today_nsq'] = pd.to_datetime (df_hist_trans['today'])

#print (df_hist_trans['today_nsq'] ,  df_hist_trans['purchase_date_nsq'])
        
df_hist_trans_group = df_hist_trans.groupby('card_id').agg(aggs)
    
pd.set_option('display.max_columns', 30)

df_hist_trans_group.columns = new_columns

type (df_hist_trans_group)

new_columns

df_hist_trans_group.columns


pd.show_versions


pd.show_versions (as_json = False)

`
******************************************************************************


Thanks

**Comment by gfyoung** [comment]
@PN12 : Thanks!  Is it also to provide the CSV files?  It's difficult to run your code otherwise.

**Comment by PN12** [comment]
@gfyoung - Please find the [link](https://drive.google.com/file/d/1zxE6DfAIdk6z7GYA4ARPLCM4owNStvC-/view?usp=sharing) for file and code.

**Comment by WillAyd** [comment]
I see that the file itself is 550 GB and that you have quite a number of aggregations you are performing, so it's not infeasible to exhaust memory on an 8 GB system. 

If you have tangible numbers and expectations around what that should be and feel that pandas is grossly exceeding that please share but otherwise closing as it's not clear there is even a problem here.

Just as a tip make sure you have the appropriate dtypes defined for your data as using object dtypes will be inefficient and possibly even cause undesirable aggregations to exhaust more memory

**Comment by PN12** [comment]
@WillAyd - 
Sorry, but the 
1. The same code did ran perfect earlier on the same laptop with 8GB RAM config.
2. Plus the Memory Error encountered above has been faced while performing pd.read_csv on the shared historical_transactions.csv file which is 2.65 GB. This operation was performed before the aggregations part. 
3. The issue I would like to understand is while the similar size files have been working fine in my data modelling projects. Then facing MemoryError multiple time for a pd.read_csv command is not understandable.  While the same file works fine on re-installing Anaconda or restarting Spyder altogether. 
4. Also, 550MB file size has worked well till now on spyder and Jupyter both. 

Would be helpful if you can guide us to the root cause and thus we can rectify it .

A re-opening of the issue would be of help as well here.

Thanks.

**Comment by PN12** [comment]
@gfyoung - Please let us know if you have inputs on the above as well which could help rectify the issue.

Thanks.

**Comment by WillAyd** [comment]
> 2\. Plus the Memory Error encountered above has been faced while performing pd.read_csv on the shared historical_transactions.csv file which is 2.65 GB. This operation was performed before the aggregations part.

If this is the case then please make your example as *minimal* as possible to reproduce the issue and provide versions where it was working before but where you feel there was now a regression.

There is unfortunately way too much unnecessary code and a variety of confounding factors being provided as is.

**Comment by PN12** [comment]
If this is the case then please make your example as minimal as possible to reproduce the issue and provide versions where it was working before but where you feel there was now a regression.

@WillAyd - Please advise on the required information and I shall share the same.

Thanks.

**Comment by WillAyd** [comment]
Check out our contributing guide as a first step:

https://pandas.pydata.org/pandas-docs/stable/contributing.html#bug-reports-and-enhancement-requests

**Comment by TomAugspurger** [comment]
@PN12 could you try

1. That example isn't quite minimal :) See http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports for some advice. For example, I don't have a C drive on my computer :) If truly necessary, you can depend external file, but that is rarely necessary. You may be able to generate the data (possibly in batches) and write it to disk, before getting a memory error reading it in. I assume scikit-learn, matplotlib, seaboard, lightgbm, etc. aren't necessary either.
2. Could you collect the requested information (minimal example, output of show_versions)
3. Nitpick: Could you format your posts (https://guides.github.com/features/mastering-markdown)? It makes them easier to scan through.

We understand that it can take some time to put together a good, minimal example. But it's what lets us figure out the problem (if any) and a solution more quickly.

**Comment by WillAyd** [comment]
Closing due to age. If you can provide info as requested above feel free to reopen
